### ISSUE TITLE
undefined name idx_token in selection clause error in db_log

### ISSUE THREAD

**Issue by aleksbykov** [scylladb/scylla#4744]
*Installation details*
Test Scylla version (or git commit hash): 3.1.rc2
Oracle Scylla version: 3.0.6
Test Cluster size: 3 nodes,
Oracle: 1 node
OS (RHEL/CentOS/Ubuntu/AWS AMI):

Run on AWS, used amis
Test cluster: 'ami-087e402340465a532'
Oracle cluster: 'ami-03415e87997ef789b'

During gemini test run, 
`gemini -d --duration 60m --warmup 10m -c 100 -m mixed -f --non-interactive --cql-features normal --max-columns 100 --min-columns 60 --max-clustering-keys 16 --max-partition-keys 8 --max-mutation-retries 5 --max-mutation-retries-backoff 100ms --test-cluster=172.30.0.231 --oracle-cluster=172.30.0.129 --outfile /tmp/gemini-l0-3c758b48-2e0f-4eca-a752-49afeeb0eb6f.log --seed 8`
on oracle cluster node appeared next errors in db log:
` [shard 2] storage_proxy - exception during mutation write to 172.30.0.129: exceptions::invalid_request_exception (Undefined name idx_token in selection clause)
`
gemini log contains next error log:
{"L":"INFO","T":"2019-07-21T06:58:46.448Z","N":"delegating_store","M":"oracle failed mutation, transition to next state impossible so continuing with next mutation","error":"unable to apply mutations to the oracle store: [cluster = oracle, query = 'INSERT INTO ks1.table1 (pk0,pk1,pk2,pk3,pk4,ck0,ck1,ck2,ck3,ck4,ck5,ck6,ck7,ck8,ck9,ck10,ck11,ck12,ck13,col0,col1,col2,col3,col4,col5,col6,col7,col8,col9,col10,col11,col12,col13,col14,col15,col16,col17,col18,col19,col20,col21,col22,col23,col24,col25,col26,col27,col28,col29,col30,col31,col32,col33,col34,col35,col36,col37,col38,col39,col40,col41,col42,col43,col44,col45,col46,col47,col48,col49,col50,col51,col52,col53,col54,col55,col56,col57) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?) ']: Operation failed for ks1.table1 - received 0 responses and 1 failures from 1 CL=QUORUM.","errorVerbose":"Operation failed for ks1.table1 - received 0 responses and 1 failures from 1 CL=QUORUM.\n[cluster = oracle, query = 'INSERT INTO ks1.table1 (pk0,pk1,pk2,pk3,pk4,ck0,ck1,ck2,ck3,ck4,ck5,ck6,ck7,ck8,ck9,ck10,ck11,ck12,ck13,col0,col1,col2,col3,col4,col5,col6,col7,col8,col9,col10,col11,col12,col13,col14,col15,col16,col17,col18,col19,col20,col21,col22,col23,col24,col25,col26,col27,col28,col29,col30,col31,col32,col33,col34,col35,col36,col37,col38,col39,col40,col41,col42,col43,col44,col45,col46,col47,col48,col49,col50,col51,col52,col53,col54,col55,col56,col57) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?) ']\ngithub.com/scylladb/gemini/store.(*cqlStore).doMutate\n\t/home/henrik/dev/gemini/store/cqlstore.go:72\ngithub.com/scylladb/gemini/store.(*cqlStore).mutate\n\t/home/henrik/dev/gemini/store/cqlstore.go:34\ngithub.com/scylladb/gemini/store.mutate\n\t/home/henrik/dev/gemini/store/store.go:97\ngithub.com/scylladb/gemini/store.delegatingStore.Mutate\n\t/home/henrik/dev/gemini/store/store.go:88\nmain.mutation\n\t/home/henrik/dev/gemini/cmd/gemini/jobs.go:189\nmain.WarmupJob\n\t/home/henrik/dev/gemini/cmd/gemini/jobs.go:90\nruntime.goexit\n\t/usr/lib/go/src/runtime/asm_amd64.s:1337\nunable to apply mutations to the oracle store\ngithub.com/scylladb/gemini/store.mutate\n\t/home/henrik/dev/gemini/store/store.go:98\ngithub.com/scylladb/gemini/store.delegatingStore.Mutate\n\t/home/henrik/dev/gemini/store/store.go:88\nmain.mutation\n\t/home/henrik/dev/gemini/cmd/gemini/jobs.go:189\nmain.WarmupJob\n\t/home/henrik/dev/gemini/cmd/gemini/jobs.go:90\nruntime.goexit\n\t/usr/lib/go/src/runtime/asm_amd64.s:1337"}

**Comment by aleksbykov** [comment]
repeated on scylla 3.0.8 as oracle instances

**Comment by nyh** [comment]
Interesting. "idx_token" is a special column that secondary index code adds to the materialized-view used to store the index, but it doesn't exist in the base table and the user shouldn't know of its existance. Do I understand correctly that the table you're updating has a secondary index enabled?

Updates to the base table cause updates to the view table, which need to fill "idx_token". Updating a view involves a select on the base table, but as far as I know we have code that makes sure never to try to select this "idx_token" column - but the error you saw suggests it was SELECTed anyway in our internal code during the write, which is a bug.

@psarna any ideas? Maybe we have a bug in this in the 3.0 branch?

@aleksbykov do you know if this error happens on the first write of the test - maybe it's some sort of race between adding an index and writing or something like that - or did it fail in the middle of a test that previously had writes succeeding?

**Comment by aleksbykov** [comment]
@nyh it looks like this happens at the beginning of test. while warmup stage is running. on this stage only write operations are performed.

**Comment by psarna** [comment]
On Sun, Jul 21, 2019, 17:46 nyh <notifications@github.com> wrote:

> Interesting. "idx_token" is a special column that secondary index code
> adds to the materialized-view used to store the index, but it doesn't exist
> in the base table and the user shouldn't know of its existance. Do I
> understand correctly that the table you're updating has a secondary index
> enabled?
>
> Updates to the base table cause updates to the view table, which need to
> fill "idx_token". Updating a view involves a select on the base table, but
> as far as I know we have code that makes sure never to try to select this
> "idx_token" column - but the error you saw suggests it was SELECTed anyway
> in our internal code during the write, which is a bug.
>
> @psarna <https://github.com/psarna> any ideas? Maybe we have a bug in
> this in the 3.0 branch?
>
Yeah, I also mentioned in on slack, I remember that it was a backporting
issue. Tomorrow, when in the office, I'll try to dig up the missing commit
and post it here.


@aleksbykov <https://github.com/aleksbykov> do you know if this error

**Comment by penberg** [comment]
@nyh @gleb-cloudius @psarna The error message is also quite confusing: error in selection clause in mutation write path -- what does that even mean?

Also, looks like these errors are not propagated to the clients:

https://github.com/scylladb/scylla/blob/master/service/storage_proxy.cc#L1629

Why is that?

**Comment by psarna** [comment]
@penberg that's because materialized views currently violate the layering a little bit by using a `select clause` in order to store information on which columns should be used in a view (grep for `select_clause` in `view.cc` for reference). It's been discussed several times that it's better to get rid of it and just keep raw schema information, but it's simply a very low priority item.

**Comment by nyh** [comment]
On Sun, Jul 21, 2019 at 7:23 PM Pekka Enberg <notifications@github.com>
wrote:

> @nyh <https://github.com/nyh> @gleb-cloudius
> <https://github.com/gleb-cloudius> @psarna <https://github.com/psarna>
> The error message is also quite confusing: error in selection clause in
> mutation write path -- what does that even mean?
>
I think you know what it means (you wrote the initial version of the
secondary indexing code :-)), but you're right that it might be confusing
to those who don't.

That being said, the error is exactly what it said: during a *write*, the
code did some *select* and failed. Clearly, the user would not know why the
heck did a write did a select, or what the heck is "token_idx" - which
isn't one of the user's own columns.  But what better can we tell the user?
We can also have more traditional read errors during the write's
read-before-write and generate those errors. I'm not sure how we can better
explain those read-errors during writes to a user who doesn't even know
that writing to a table with a secondary index involves also an
internally-generated read.

Do you have a better phrase you would have used?


> Also, looks like these errors are not propagated to the clients:
>
>
> https://github.com/scylladb/scylla/blob/master/service/storage_proxy.cc#L1629
>
> Why is that?
>
I don't have time right now to look into the entire path here, you may be
right. But note that in general, Cassandra's MV CL support only tells the
client when enough base replicas - not view replicas - have been written.
But you're right that if we have a problem with the read-before write and
know that we're going to fail all view writes even before doing the base
write, we should probably fail the base write as well instead of pretending
it succeeded. So that's a good point that probably deserves its own issue
(although maybe I'm missing something here because I didn't have time to
look into the code).

**Comment by gleb-cloudius** [comment]
On Sun, Jul 21, 2019 at 09:23:25AM -0700, Pekka Enberg wrote:
> Also, looks like these errors are not propagated to the clients:
> 
> https://github.com/scylladb/scylla/blob/master/service/storage_proxy.cc#L1629
> 
> Why is that?
> 
Why should it? A write operation consists of multiple replica writes. If
one fails (as it the case here) it does not mean entire write will. If
enough individual replica write fail so that achieving CL is no longer
possible the operation will fail and error will be sent to a client with
info about how much individual operation failed.

--
			Gleb.

**Comment by psarna** [comment]
https://github.com/scylladb/scylla/issues/4478 is the issue, https://github.com/scylladb/scylla/commit/e438e3e3ce41f6c878b111f5c19bf2b28aa51098 should be backported.

**Comment by slivne** [comment]
So do we need to apply this patch also on 3.0 branch it seems to be in 3.1 ?

**Comment by penberg** [comment]
@slivne @psarna Something is not right...

We hit the issue in the _oracle_ node, which is running 3.0.6 and @aleksbykov managed to trigger the bug with 3.0.8 also.

However, the broken commit dbe8491655a3ffb6ce00c66a43a74bb2c5732382 ("view: cache is_index for view pointer") is not in `branch-3.0` so there's no need for the revert.

**Comment by psarna** [comment]
Hm, if it's not a race between caching `is_index` and registering an index, then the only other thing I can think of is a race in caching the `_select_statement`. I don't know *how* `select_statement()` ends up called before the view is registed as being an index, but it would certainly result in the error we're seeing here.

Bright news is that after the `computed_columns` series, which is in, will not have that problem anyway, because then we no longer care if a view is an index or not.

For now, one remedy would be to recompute `select_statement` every time, but that's a big waste of CPU, because it's frequently used in partition_slice(), partition_ranges(), etc. Perhaps it's still negligible though, I haven't profiled it.

The other way would be to pinpoint where exactly this race happens and try to avoid it (perhaps we need to wait for schema sync somewhere? no idea).


Also, it would be nice to stop using `select_statement` here, because it doesn't really belong. What we need is `partition_slice`, `partition_ranges` and a couple of restrictions, but these we can store directly.

**Comment by juliayakovlev** [comment]
Same problem receive in the MV&SI longevity.
Scylla version 3.1.0.rc9-0.20191006.05272c53e

```
< t:2019-10-06 22:32:42,964 f:cluster.py      l:1231 c:sdcm.cluster         p:DEBUG > 2019-10-06T22:32:39+00:00  ip-10-0-171-238 !INFO    | scylla: [shard 7] schema_tables - Altering sec_index.users id=89f4bee0-e840-11e9-bd7b-00000000000
a version=ae4fb57f-337d-3737-8fba-319a98b9bd86

< t:2019-10-06 22:32:42,964 f:cluster.py      l:1231 c:sdcm.cluster         p:DEBUG > 2019-10-06T22:32:39+00:00  ip-10-0-171-238 !INFO    | scylla: [shard 7] schema_tables - Creating sec_index.users_address_ind_index id=33e85ae1-e889-11e
9-8f7b-000000000000 version=d276cca4-7303-3f47-9157-864aab9f12e8

< t:2019-10-06 22:32:42,964 f:cluster.py      l:1231 c:sdcm.cluster         p:DEBUG > 2019-10-06T22:32:39+00:00  ip-10-0-171-238 !INFO    | scylla: [shard 10] query_processor - Column definitions for sec_index.users changed, invalidating
 related prepared statements

< t:2019-10-06 22:32:42,964 f:cluster.py      l:1231 c:sdcm.cluster         p:DEBUG > 2019-10-06T22:32:39+00:00  ip-10-0-171-238 !ERR     | scylla: [shard 7] storage_proxy - exception during mutation write to 10.0.171.238: exceptions::invalid_request_exception (Undefined name idx_token in selection clause)

< t:2019-10-06 22:32:42,965 f:cluster.py      l:1231 c:sdcm.cluster         p:DEBUG > 2019-10-06T22:32:39+00:00  ip-10-0-171-238 !INFO    | scylla: [shard 9] query_processor - Column definitions for sec_index.users changed, invalidating related prepared statements

< t:2019-10-06 22:32:42,965 f:cluster.py      l:1231 c:sdcm.cluster         p:DEBUG > 2019-10-06T22:32:39+00:00  ip-10-0-171-238 !INFO    | scylla: [shard 3] query_processor - Column definitions for sec_index.users changed, invalidating related prepared statements

< t:2019-10-06 22:32:42,965 f:cluster.py      l:1231 c:sdcm.cluster         p:DEBUG > 2019-10-06T22:32:39+00:00  ip-10-0-171-238 !INFO    | scylla: [shard 5] query_processor - Column definitions for sec_index.users changed, invalidating related prepared statements

< t:2019-10-06 22:32:42,966 f:cluster.py      l:1231 c:sdcm.cluster         p:DEBUG > 2019-10-06T22:32:39+00:00  ip-10-0-171-238 !WARNING | scylla: [shard 3] storage_proxy - Failed to apply mutation from 10.0.166.201#3: exceptions::invalid_request_exception (Undefined name idx_token in selection clause)
```

**Comment by psarna** [comment]
This issue is effectively fixed on master with db8d4a0c, but it's a feature series, not a fix, so it's not eligible for backporting. The only suspect I have is a race between calling `get_view_row` and a schema update, which works differently on master, because the info if a column is computed is kept in the `system_schema` tables. In 3.1 it's not the case yet - instead we check if the view we're using actually backs a secondary index, so the only possible explanation I can come up with is some kind of a race - e.g. the node hasn't realised yet that a view is actually used by an index, and at the same time we try to generate a view update and fail because an unknown "idx_token" column appears. However, I wasn't able to trigger this hypothetical race even once last time I tried.

I suspect that if it is indeed such a race, the right approach for fixing it should actually be similar to what's done in db8d4a0c - marking the column as being `computed` in the schema info itself, so this whole check can be index-agnostic - we don't care if the view backs an index and we don't need to check it anywhere, because the column itself is marked as a special column with values generated from other columns. This change is however too big a fix to simply be backported, in my opinion. That's because it involved adding a new `schema_feature` and many other fragile backward compatibility issues. A quick&dirty workaround would be a patch that specifically catches this exception and lets it continue if the unknown column happens to be called `idx_token`, but it's a dirty, dirty hack, not a solution. The proper way is already merged to master.

**Comment by penberg** [comment]
@psarna I guess we can close this issue then?

**Comment by psarna** [comment]
yup

**Comment by penberg** [comment]
@avikivity @slivne We're hitting this bug in Gemini testing, and no stable version carries the fix. Shoud we backport the computed columns series to 3.1? @psarna, how risky do you think that is?

**Comment by psarna** [comment]
4 out of 5, computed column is a feature in all meanings of this word - it also has its own cluster feature and a schema feature as it adds a table to `system_schema`. I'd say it's a big no-no.

**Comment by avikivity** [comment]
@nyh already backported this to 3.2, 3.3. 3.1 turned out not to be vulnerable.

**Comment by avikivity** [comment]
@nyh says this might not be fixed by the patch.

**Comment by nyh** [comment]
The mentioned patch fixes code which was introduced in 3.2, and wasn't present in 3.1, which was the version against which this bug was reported. 
@aleksbykov @penberg could you please try reproducing this issue on master, to see if it still exists or not?

**Comment by psarna** [comment]
To be strict, it was fixed for 3.2 by computed columns, but I think 9b379e3 is still related if the test was rerun with mixed 3.2 and 3.1 nodes in the same cluster.

**Comment by avikivity** [comment]
What about 3.1? The issue say the bug was discovered on 3.1.rc2.

**Comment by psarna** [comment]
Regarding 3.1 I only have educated guesses - https://github.com/scylladb/scylla/issues/4744#issuecomment-513758812 and https://github.com/scylladb/scylla/issues/4744#issuecomment-545958611. The fact is that the presence of computed columns eliminates this branch, because instead of guessing which columns are probably computed, we'd simply know what, because the information will exist inside the schema.

**Comment by tzach** [comment]
@psarna @avikivity can we close this? The patch is part of 3.2.1

**Comment by slivne** [comment]
the issue is fixed in 3.2,3.3

3.1 is not clear - we werent able to reproduce

**Comment by avikivity** [comment]
Already backported to all vulnerable branches, removing "Backport candidate" label.
